pkg/query/query_bench_test.go

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package query
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/go-kit/kit/log"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/thanos-io/thanos/pkg/gate"
+	"github.com/thanos-io/thanos/pkg/store/labelpb"
+	"github.com/thanos-io/thanos/pkg/store/storepb"
+	storetestutil "github.com/thanos-io/thanos/pkg/store/storepb/testutil"
+	"github.com/thanos-io/thanos/pkg/testutil"
+)
+
+// TestQuerySelect benchmarks querier Select method. Note that this is what PromQL is using, but PromQL might invoke
+// this many times and within different interval e.g
+// TODO(bwplotka): Add benchmarks with PromQL involvement.
+func TestQuerySelect(t *testing.T) {
+	tb := testutil.NewTB(t)
+	storetestutil.RunSeriesInterestingCases(tb, 200e3, 200e3, func(t testutil.TB, samplesPerSeries, series int) {
+		benchQuerySelect(t, samplesPerSeries, series, true)
+	})
+}
+
+func BenchmarkQuerySelect(b *testing.B) {
+	tb := testutil.NewTB(b)
+	storetestutil.RunSeriesInterestingCases(tb, 10e6, 10e5, func(t testutil.TB, samplesPerSeries, series int) {
+		benchQuerySelect(t, samplesPerSeries, series, true)
+	})
+}
+
+func benchQuerySelect(t testutil.TB, totalSamples, totalSeries int, dedup bool) {
+	tmpDir, err := ioutil.TempDir("", "testorbench-queryselect")
+	testutil.Ok(t, err)
+	defer func() { testutil.Ok(t, os.RemoveAll(tmpDir)) }()
+
+	const numOfReplicas = 2
+
+	samplesPerSeriesPerReplica := totalSamples / numOfReplicas
+	if samplesPerSeriesPerReplica == 0 {
+		samplesPerSeriesPerReplica = 1
+	}
+	seriesPerReplica := totalSeries / numOfReplicas
+	if seriesPerReplica == 0 {
+		seriesPerReplica = 1
+	}
+
+	random := rand.New(rand.NewSource(120))
+	var resps []*storepb.SeriesResponse
+	var expectedSeries []labels.Labels
+	for j := 0; j < numOfReplicas; j++ {
+		// Note 0 argument - this is because we want to have two replicas for the same time duration.
+		head, created := storetestutil.CreateHeadWithSeries(t, 0, storetestutil.HeadGenOptions{
+			TSDBDir:          filepath.Join(tmpDir, fmt.Sprintf("%d", j)),
+			SamplesPerSeries: samplesPerSeriesPerReplica,
+			Series:           seriesPerReplica,
+			Random:           random,
+			PrependLabels:    labels.FromStrings("a_replica", fmt.Sprintf("%d", j)), // a_ prefix so we keep sorted order.
+		})
+		testutil.Ok(t, head.Close())
+		for i := 0; i < len(created); i++ {
```

Comment:
This whole section is difficult parse what is going on and why - care to add a comment?